model/map-file/src/main/java/org/keycloak/models/map/storage/file/FileMapKeycloakTransaction.java

```diff
@@ -0,0 +1,140 @@
+import org.keycloak.models.map.storage.chm.MapFieldPredicates;
+import org.keycloak.models.map.storage.chm.MapModelCriteriaBuilder.UpdatePredicatesFunc;
+import org.keycloak.storage.SearchableModelField;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.util.IdentityHashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * {@link MapKeycloakTransaction} implementation used with the file map storage.
+ *
+ * @author <a href="mailto:sguilhen@redhat.com">Stefan Guilhen</a>
+ */
+public class FileMapKeycloakTransaction<V extends AbstractEntity & UpdatableEntity, M>
+  extends ConcurrentHashMapKeycloakTransaction<String, V, M> {
+
+    private final List<Path> touchedPaths = new LinkedList<>();
+    private Map<Path, Path> renameOnCommit = new IdentityHashMap<>();
+
+    private final String txId = StringKey.INSTANCE.yieldNewUniqueKey();
+
+    public static <V extends AbstractEntity & UpdatableEntity, M> FileMapKeycloakTransaction<V, M> newInstance(Class<V> entityClass,
+      Function<String, Path> dataDirectoryFunc, Function<V, String[]> suggestedPath,
+      boolean isExpirableEntity, Map<SearchableModelField<? super M>, UpdatePredicatesFunc<String, V, M>> fieldPredicates) {
+        Crud<V, M> crud = new Crud<>(entityClass, dataDirectoryFunc, suggestedPath, isExpirableEntity, fieldPredicates);
+        FileMapKeycloakTransaction<V, M> tx = new FileMapKeycloakTransaction<>(entityClass, crud);
+        crud.tx = tx;
+        return tx;
+    }
+
+    private FileMapKeycloakTransaction(Class<V> entityClass, Crud<V, M> crud) {
+        super(
+          crud,
+          StringKeyConverter.StringKey.INSTANCE,
+          DeepCloner.DUMB_CLONER,
+          MapFieldPredicates.getPredicates(ModelEntityUtil.getModelType(entityClass)),
+          ModelEntityUtil.getRealmIdField(entityClass)
+        );
+    }
+
+    @Override
+    public void rollback() {
+        this.touchedPaths.forEach(FileMapKeycloakTransaction::silentDelete);
+        super.rollback();
+    }
+
+    @Override
+    public void commit() {
+        super.commit();
+        this.renameOnCommit.forEach(FileMapKeycloakTransaction::silentMove);
+        this.touchedPaths.forEach(FileMapKeycloakTransaction::silentDelete);
+    }
+
+    private static void silentMove(Path from, Path to) {
+        try {
+            Files.move(from, to, StandardCopyOption.REPLACE_EXISTING);
+        } catch (IOException ex) {
+            throw new UncheckedIOException(ex);
+        }
+    }
+
+    private static void silentDelete(Path p) {
+        try {
+            if (Files.exists(p) && Files.size(p) == 0) {
+                Files.delete(p);
+            }
+        } catch (IOException e) {
+            // Swallow the exception
```

Comment:
This is _silent_ delete and the nonexistence of the file is expected, thus no exception should be thrown.

Reply:
**TL;DR**: Doing some search where this is being called from and when, this seems mostly harmless. Still, this swallowed exception will never occur when the file has already been deleted. 

In a previous review I asked for some comments when then this method should be used. Tracing this back it seems only be used for `touchedPaths`. Looking where this is being filled, it seems to used for temporary files. Looking at the body of the method, it also handles only empty files. It also checks if the file exists. 

A deletion of a file might fail due to multiple reasons, not only because it no longer exists. The file system might be a network file system and the operation might temporarily fail. Other reasons would be that the file system might be mounted read-only, the Keycloak instance might lack the required permissions (unlikely here as Keycloak created the files earlier in the transaction).  In such a situation, when the file exists, but can't be deleted, throwing an exception is IMHO valid. Not sure when those temporary files would cause problems later on. 

So to summarize, looking at the code, there should be no `IOException` when the file doesn't exist as it is handled via `Files.exists(p)`. There might be other `IOException`s for other reasons as described above, although I don't know if those would be harmful here / would lead to problems eventually when the file is not removed. 

So it is still not obvious to me why it is either important to swallow the exception here. 

The case where a delete that isn't executed is problematic is then probably here in `FileMapStoreage.Crud#delete()` where there is currently only warning.